lib/ansible/module_utils/docker_common.py

```diff
@@ -0,0 +1,498 @@
+                 required_together=None, required_if=None):
+
+        self.logger = logging.getLogger(self.__class__.__name__)
+
+        merged_arg_spec = dict()
+        merged_arg_spec.update(DOCKER_COMMON_ARGS)
+        if argument_spec:
+            merged_arg_spec.update(argument_spec)
+            self.arg_spec = merged_arg_spec
+
+        mutually_exclusive_params = []
+        mutually_exclusive_params += DOCKER_MUTUALLY_EXCLUSIVE
+        if mutually_exclusive:
+            mutually_exclusive_params += mutually_exclusive
+
+        required_together_params = []
+        required_together_params += DOCKER_REQUIRED_TOGETHER
+        if required_together:
+            required_together_params += required_together
+
+        self.module = AnsibleModule(
+            argument_spec=merged_arg_spec,
+            supports_check_mode=supports_check_mode,
+            mutually_exclusive=mutually_exclusive_params,
+            required_together=required_together_params,
+            required_if=required_if)
+
+        if not HAS_DOCKER_PY:
+            self.fail("Failed to import docker-py - %s. Try `pip install docker-py`" % HAS_DOCKER_ERROR)
+
+        if docker_version < MIN_DOCKER_VERSION:
+            self.fail("Error: docker-py version is %s. Minimum version required is %s." % (docker_version,
+                                                                                           MIN_DOCKER_VERSION))
+
+        debug = self.module.params.get('debug')
+        log_mode = self.module.params.get('log_mode')
+        filter_logger = self.module.params.get('filter_logger')
+        if debug and log_mode == 'syslog':
+            handler = DockerSysLogHandler(level=logging.DEBUG, module=self.module)
+            self.logger.addHandler(handler)
+            logging.basicConfig(level=logging.DEBUG)
+        elif debug and log_mode == 'file':
+            log_path = self.module.params.get('log_path')
+            logging.basicConfig(level=logging.DEBUG, filename=log_path)
+        elif debug and log_mode == 'stderr':
+            logging.basicConfig(level=logging.DEBUG, stream=sys.stderr)
+
+        if filter_logger:
+            for h in logging.root.handlers:
+                h.addFilter(logging.Filter(name=self.logger.name))
+
+        self.check_mode = self.module.check_mode   
+        self._connect_params = self._get_connect_params()
+
+        try:
+            super(AnsibleDockerClient, self).__init__(**self._connect_params)
+        except APIError, exc:
+            self.fail("Docker API error: %s" % exc)
+        except Exception, exc:
+            self.fail("Error connecting: %s" % exc)
+
+    def log(self, msg, pretty_print=False):
+        self.logger.debug(log_msg(msg, pretty_print=pretty_print))
+    
+    def fail(self, msg):
+        self.module.fail_json(msg=msg)
+
+    @staticmethod
+    def _get_value(param_name, param_value, env_variable, default_value):
```

Comment:
pretty sure this specific to docker_common. especially where we drop in specific values for cert_path, cacert_path and key_path.